OpenGL: Add Palette-Perfect Transparency

LRFLEW opened the pull request:
This is based off of my work on #6298. After @duncanspumpkin raised concerns about merging it without a proper solution to sprite transparency, I decided to work on adding proper transparency for all objects using the OpenGL renderer. The process of making this work was long, and was filled with many failed experiments. Eventually, I found a fast solution that renders transparency pixel-perfect with the software renderer.

There were a lot of changes made here, so I'll try to break down the major points, and explain why I made the changes.

Starting with the first commit:

* **#6298:** This includes my previous PR, where I converted the Framebuffers to work in palette space instead of RGB space. This was a key step in making the transparency work efficiently. This removed the for-loop in the old FillRect shader, which dramatically improved performance.

* **A quick word on #6465:** I made that test as I was deciding whether or not to make that change here. While using Geometry Shaders would *technically* decrease the minimum OpenGL requirement (from 3.3 to 3.2), I decided not to go with that change. Looking at OpenGL ES (which I want to allow this to use at some point), Geometry Shaders were added in ES 3.2 (a very new standard), whereas the instancing functions all are in ES 3.0. To make future OpenGL ES support easier, I left the instancing code as-is.

Moving on to the shader improvements:

**Note:** I removed all transparent object rendering in this commit. This was to make the optimizations easier. I worked without the transparent objects, then added them back in later.

* **Moved Clipping from fragment shader to vertex shader:** This one is pretty simple. By trimming the primitive in the vertex shader, we avoid unnecessary iterations of the fragment shader. I saw a significant performance boost from this change.

* **Handled remaps more directly in the shader:** This started with uploading all the palette remaps to their own texture (a step needed later for transparency anyways). With this palette, I removed all the palettes from the Sprite Cache. The multiple remappings are handled in the shader, instead of creating thousands of palettes for each combination. This seemed to help reduce the stuttering caused by updating the Sprite Cache a significant amount.

* **Removed vIndex, and used matrices:** To allow us to send the size and location of the stuff to render only once, we had one static vertex input, which was an index 0-3, specifying which corner was being drawn. This was an inelegant solution, so I replaced it with a matrix multiplication. This was mostly to make the vertex shader code cleaner, but it had a very small performance improvement for me.

* **Merged DrawImage and FillRect into one shader, DrawRect:** This was to make adding transparency back easier. This provided a small performance boost, due to it causes less command buffer flushes, but all the excess buffer flushes were removed in the next commit anyways.

Next was adding the Depth Buffer:

* I needed to have a depth value for the object being rendered, so I hacked one together. I added a counter that records how many objects were added before, and use that value for each object to create a depth value. At some point, this should be replaced with a depth value defined by the objects location in world space (and not screen space), but that would require reworking the IDrawingContext interface.

* The added depth value allowed me to avoid flushing the buffers until the EndDraw function, and add instancing to the DrawLine shader. This had a small bump in performance, though with having already merged DrawImage and FillRect, the buffers weren't getting flushed that often anyways.

Next was the single-pass transparency:

* This renders the transparent objects to a separate Framebuffer using the DrawRect shader. Being able to reuse the DrawRect shader made this easier. The value being written to the Framebuffer is the y index for the palette in the palette texture. This is then combined with the main framebuffer using a new shader that applies the palette remapping.

* This uses the depth buffer from before to get the top layer of transparency. This depth buffer is also used with the original depth buffer to make sure transparency isn't applied to objects in front of the transparent object.

The idea of only handling one transparency layer was based on a suggestion from @duncanspumpkin on #6298. However, seeing it in action shows that it isn't a great solution. For one, the transparent UI elements (like the main menu buttons, bottom toolbar, and top toolbar menus) are rendered with two transparency passes. The first one adds a translucency effect, while the second adds the color. Even if we made a special case for this example, there are in-world transparent objects that are visually changed by transparent objects behind it. For example, light blue glass in front of black glass is noticeably different from either glass colors alone.

This meant that multiple transparency passes were clearly necessary. Figuring out how to do it was a challenge. The problem is that it's time-expensive to rely on the results of a previous draw, so swapping the buffers for every transparent object isn't a solution. The challenge then, is to minimize the number of buffers swaps, while still rendering all of the transparency layers accurately. This means finding groups of non-overlapping transparent objects to render at once.

One major problem with checking for the overlapping is the tiling done for sprites like water. If you just check the bounding box, then you'll get a huge amount of false positives. My first attempt at avoiding this, involving R-Trees (a data structure similar to Quad-Trees) and multiple transparency buffers, failed to produce the correct drawing order. I had made a comment in #6298 about basically running a variation of the software renderer to find non-overlapping objects, but while reconsidering it, I came up was a better solution.

* **Depth Peeling:** First documented by [Nvidia](http://developer.download.nvidia.com/assets/gamedev/docs/OrderIndependentTransparency.pdf), this allows you to use the GPU to produce the multiple layers of transparent objects. Put simply, you use the depth buffer from the previous transparency slice to exclude fragments that have already been rendered. Now this implementation works a bit differently, going back-to-front to be able to properly handle the palette remaps.

There is a problem with this: the technique doesn't have a way of knowing how many layers are needed. The Nvidia paper recommended using a fixed number of iterations, but the back-to-front nature of this implementation made that unviable. Clearly, we can stop when the last transparency layer is empty, but it's hard to get that information from the GPU. I tried a number of methods, but they all ran very slowly for me (30fps or slower). (You can check out my attempts [in the commits on my other branch](https://github.com/LRFLEW/OpenRCT2/commits/opengl-working))

One advantage we have is that we don't need to know the exact number of iterations needed. If we perform too many iterations, then the performance might be a little worse, but there is no visual change to the rendering. This means that a test with the bounding boxes can produce a reasonable approximation for the number of iterations needed.

* **Sweepline Algorithm:** Based on a suggestion from @AaronVanGeffen, I wrote a sweep line algorithm to find the maximum number of overlapping objects based on their bounding boxes. This doesn't necessarily produce the exact number of iterations needed, but my testing indicates that it's only off by a few. This is the best compromise between accuracy and speed I could find.

In the end, this ended up performing rather well from my tests. Adding in the multi-pass transparency did hurt the framerate in my tests, but the other improvements I made along the way were able to increase the framerate more than the transparency was able to take away in most situations. All of my tests did show this OpenGL renderer being faster than the Software renderer at least.

I'm sure I need to fix some code style mistakes, and I want to add comments to the sweepline algorithm, but this should be functionally complete.

Gymnasiast commented on src/openrct2-ui/drawing/engines/opengl/ApplyTransparencyShader.cpp:
```diff
@@ -0,0 +1,94 @@
+
+#include "ApplyTransparencyShader.h"
+
+struct VDStruct {
```
Brace needs to go on its own line.

LRFLEW replied:
I tried to remember that for the if/else/while/for, but forgot it here. I'll fix it.

Gymnasiast commented on src/openrct2-ui/drawing/engines/opengl/CopyFramebufferShader.cpp:
```diff
@@ -18,23 +18,37 @@
 
 #include "CopyFramebufferShader.h"
 
+struct VDStruct {
```
Ditto

Gymnasiast commented on src/openrct2-ui/drawing/engines/opengl/CopyFramebufferShader.cpp:
```diff
@@ -46,38 +60,26 @@ CopyFramebufferShader::~CopyFramebufferShader()
     OpenGLAPI::SetTexture(0, GL_TEXTURE_2D, texture);
 }
 
+void CopyFramebufferShader::SetPalette(const vec4 * glPalette) {
```
Ditto

Gymnasiast commented on src/openrct2-ui/drawing/engines/opengl/DrawCommands.h:
```diff
@@ -50,57 +50,64 @@ class CommandBatch
-    vec4f colour;
-    vec4i bounds;
-    sint32 mask;
+struct DrawRectCommand {
```
ditto

Gymnasiast commented on src/openrct2-ui/drawing/engines/opengl/DrawLineShader.cpp:
```diff
@@ -19,20 +19,53 @@
 #include "DrawLineShader.h"
 #include "OpenGLFramebuffer.h"
 
+struct VDStruct {
```
ditto

Gymnasiast commented on src/openrct2-ui/drawing/engines/opengl/DrawRectShader.cpp:
```diff
@@ -0,0 +1,164 @@
+
+#include "DrawRectShader.h"
+
+struct VDStruct {
```
ditto

Gymnasiast commented on src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp:
```diff
@@ -80,11 +82,13 @@ class OpenGLDrawingContext final : public IDrawingContext
+    
+    sint32 _drawCount  = 0;
 
     struct {
```
ditto

Gymnasiast commented on src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp:
```diff
@@ -80,11 +82,13 @@ class OpenGLDrawingContext final : public IDrawingContext
-        ImageCommandBatch images;
+        RectCommandBatch rects;
+        RectCommandBatch transparent;
     } _commandBuffers;
```
ditto

Gymnasiast commented on src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp:
```diff
@@ -908,7 +893,7 @@ void OpenGLDrawingContext::SetDPI(rct_drawpixelinfo * dpi)
     _clipLeft = (sint32)(bitsOffset % (screenDPI->width + screenDPI->pitch));
     _clipTop = (sint32)(bitsOffset / (screenDPI->width + screenDPI->pitch));
-
+    
```
trailing whitespace

Gymnasiast commented on src/openrct2-ui/drawing/engines/opengl/SwapFramebuffer.cpp:
```diff
@@ -20,49 +20,49 @@
+    glDisable(GL_DEPTH_TEST);
+    shader.Use();
+    shader.SetTextures(
+                       _opaqueFramebuffer.GetTexture(),
```
Should only be intended once.

Gymnasiast commented on src/openrct2-ui/drawing/engines/opengl/SwapFramebuffer.cpp:
```diff
@@ -20,49 +20,49 @@
+                       _transparentFramebuffer.GetTexture(),
+                       _transparentFramebuffer.GetDepthTexture(),
+                       paletteTex
+                       );
```
Should be at the same indentation level as `shader.SetTextures(`.

Gymnasiast commented:
There are only minor formatting things, you did pretty well in that area. 👍

AaronVanGeffen commented:
Amazing work, @LRFLEW! This must have taken you so much work, it's very nice to see it come to fruition. I'm very excited to test it!

I do have a quick question, though. Does the framebuffer switch to palette space influence the way shaders work at all? I'm asking with @Dennis-Z's lighting/shadow experiment in mind.

duncanspumpkin commented on src/openrct2-ui/drawing/engines/opengl/TextureCache.cpp:
```diff
@@ -128,36 +87,61 @@ void TextureCache::CreateAtlasesTexture()
+    rct_drawpixelinfo dpi = CreateDPI(256, PALETTE_TO_G1_OFFSET_COUNT + 5);
+    
+    // Init no-op palette
+    for (int i=0; i < 256; ++i)
```
Why init to 0->255? Is there a reason 0 cant be used?

LRFLEW replied:
The idea here is to allow for a no-op without a branch in the shader. If you look at the ApplyTransparency shader, you can see it used there. Our use of branches in the DrawImage/DrawRect shader [is almost Undefined Behavior](https://www.khronos.org/opengl/wiki/Sampler_(GLSL)#Non-uniform_flow_control), so I added a way for producing the input value as output without requiring a branch. This isn't 100% necessary, but having it is still useful.

Plus with the water palette being 5 pixels high, adding a few more rows doesn't increase the complexity of mapping from palette index to y value all that much.

duncanspumpkin commented on src/openrct2-ui/drawing/engines/opengl/TextureCache.cpp:
```diff
@@ -128,36 +87,61 @@ void TextureCache::CreateAtlasesTexture()
-    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
-    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
-    glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
+    rct_drawpixelinfo dpi = CreateDPI(256, PALETTE_TO_G1_OFFSET_COUNT + 5);
```
Is the +5 because of water?

LRFLEW replied:
It's +1 for the no-op palette, and +(5 - 1) for the water palette.

duncanspumpkin commented:
CopyFrameBuffer needs to be renamed. It no longer copies the frame buffer but converts the palette space to RGB space.

janisozaur commented:
[circus circus adventuredome 2017-10-23 23-11-03](https://user-images.githubusercontent.com/550290/31913511-9987722e-b847-11e7-891e-95adcc94aa06.png)

Trying to get a screenshot with your OpenGL changes results in the image above.

LRFLEW commented:
To make a note in the PR, @AaronVanGeffen noted that the DPI scaling feature doesn't work right with this. Because of the integer buffers, the DPI scaling uses NN, despite whatever setting you set in the options. This can be fixed with an additional Framebuffer copy. I'll work on adding that to this PR.

janisozaur commented:
Yes, I do have scaling set to 2.

LRFLEW commented:
@janisozaur I have no idea what's causing your screenshot error, as I couldn't reproduce it.

![circus circus adventuredome 2017-10-23 16-30-34](https://user-images.githubusercontent.com/607715/31914352-f9248bf0-b80f-11e7-9ee5-b18fd3b88637.png)

There's a chance that it's a bug with whatever build/version of libpng you're using. It could also be the DPI's stride, but there should be an assert there that would trigger before that. I will keep an eye out for a possible cause, though.

LRFLEW commented:
Right after making that comment, I figured it out, and was able to reproduce it. We set the `GL_PACK_ALIGNMENT` parameter, but not the `GL_UNPACK_ALIGNMENT` parameter, so any width that's not a multiple of 4 would cause an unexpected stride. I'll push a fix.

LRFLEW commented:
While playing on Blind's server for a bit, I found a few more rendering bugs. Ignoring the ones that I know were there before (such as zooming not being quite accrue and ground transitions/dithering being slightly off), I found two that may be caused by this PR, but have no clue how it could be happening.

Disabled buttons (particularly in the ride construction window) render as boxes, instead of the proper sprite shape. I checked the code location I thought the error would be, and didn't see any problems. 
[Screenshot](https://user-images.githubusercontent.com/607715/31919961-8ce34082-b82b-11e7-965d-30ce874afb04.png)

When placing sub-voxel scenery (so half-size/quarter-size scenery), a black grid appears below the scenery ghost. Using the screenshot feature, I was able to determine that the grid is drawing the 0th palette index (so the "transparent" color). 
[Screenshot](https://user-images.githubusercontent.com/607715/31920010-cd92ee02-b82b-11e7-9a64-f4654f0beb86.png)

LRFLEW commented:
Fixed both of the issues I found. The first was caused by a mistake in `OpenGLDrawingContext::DrawSpriteSolid` based on my shader changes (solid color sprites are done using the mask attributes). The second was a mis-judge of how the palette-mapping could work, and required an additional `discard` clause.

I also removed the whitespace on the empty lines in both the source files and shader files I messed with.

LRFLEW commented:
I believe I've addressed all of the issues and bugs raised. @duncanspumpkin @Gymnasiast When you have the chance, this should be ready for another review.

@AaronVanGeffen The lighting/shadow experiment @Dennis-Z made is affected by this, but mostly by the changes carried over from #6298. In that thread, @Dennis-Z chimed in, and we figured out a way that the two features could be merged. This PR also has the shader optimizations, but that should be fairly easy to merge the shadow code into. That should be it for any possible complications. @Dennis-Z, if you could take a look over the changes and make sure that nothing overly breaks your code, that might help.

duncanspumpkin commented:
You just need to fix the asserts to make the ci builds happy and this looks good to go

Dennis-Z commented:
I think the adjustments that have to be made to that lighting prototype will be similar as discussed before through storing an extra RGB color attachment for the various framebuffer used now. Probably have to do a palette -> RGB conversion for viewing light through transparent things in the applytransparency fragment shader too. So it'll probably be fine :+1:

LRFLEW commented:
@duncanspumpkin What exactly do you mean by "fix the asserts"? It looks like the CI builds are stalling trying to pull the docker instance, which is independent of what this PR is doing.

Gymnasiast commented:
@LRFLEW Two Travis jobs failed because of network problems, but one failed because of an actual problem with the code.

duncanspumpkin commented:
@lrflew oh sorry that was what was causing them to fault before and i never actually checked to see if it was something different

janisozaur commented:
I have verified it working on my Intel GPU. I have not done any kinds of performance testing.

It doesn't look quite right with SWR or llvmpipe, but I'm not able to tell where the fault is. I'm willing to assume it's on their side, although current `develop` works fine with them. See ![openrct2_162](https://user-images.githubusercontent.com/550290/32076892-cbc9d55e-baa1-11e7-966f-121b5a55c39d.png) about what I mean. It's up to you if you want to look into that.

Travis is now green as well.

LRFLEW commented:
>It doesn't look quite right with SWR or llvmpipe, but I'm not able to tell where the fault is. I'm willing to assume it's on their side, although current develop works fine with them.

It could go either way. It's possible that I'm making use of some undefined/implementation-defined behavior I haven't caught (though if all hardware implementations work fine, it's not likely that). It's also possible that SWR/llvmpipe is out-of-standard by doing something like giving less depth-buffer precision than requested, or not correctly guaranteeing memory coherency. It's hard to tell just from the screenshot. Did you see any "OpenGL error" messages in the console output? If so, then it's likely a mistake I made. Otherwise, it's more likely an issue with their code.

If this works with all hardware implementations tested, then this should be fine to merge, as any setup not using an actual GPU should probably use the game's built-in software renderer instead 😜.

janisozaur commented:
No, I have not seen any errors being reported, not even with Mesa's `LIBGL_DEBUG=1` environment variable set. Maybe setting it to `verbose` would shed some more light.

But just like said – having it working with actual GPUs is what is important and it looks like it was tested a bit. It would be still nice to see it tested with non-Intel, non-Nvidia GPU.

LRFLEW commented:
> It would be still nice to see it tested with non-Intel, non-Nvidia GPU.

@spacek531 You tested #6465 with an AMD APU. Mind spinning this up and making sure it looks (and runs) ok? ([here's the AppVeyor build if you need it](https://ci.appveyor.com/api/buildjobs/w51kpv5g8ap7ti3p/artifacts/artifacts%2Fopenrct2-portable-release-x64-83a6543.zip))

duncanspumpkin commented:
I treated it on an amd gpu worked fine no issues.

janisozaur commented:
I couldn't find any reports even with `LIBGL_DEBUG=verbose`, the highest there is. If I do, I will report them.

LGTM.

janisozaur commented:
Oh, I just noticed – it requires some new files, have you checked they all get packaged and installed properly outside of dev environment / on CIs?

LRFLEW commented:
Um... I haven't actually checked. As I understand it, the Windows build just copies the entire data directory, so that should be fine. (I can at least confirm that the AppVeyor artifacts have all the new files). I don't know about the CMake build, and don't understand it well enough to check it myself. Mind taking a look at that for me?

janisozaur commented:
```
OpenRCT2
├── data
│   ├── g2.dat
│   ├── language
│   │   ├── ar-EG.txt
│   │   ├── ca-ES.txt
│   │   ├── cs-CZ.txt
│   │   ├── de-DE.txt
│   │   ├── en-GB.txt
│   │   ├── en-US.txt
│   │   ├── es-ES.txt
│   │   ├── fi-FI.txt
│   │   ├── fr-FR.txt
│   │   ├── hu-HU.txt
│   │   ├── it-IT.txt
│   │   ├── ja-JP.txt
│   │   ├── ko-KR.txt
│   │   ├── nb-NO.txt
│   │   ├── nl-NL.txt
│   │   ├── pl-PL.txt
│   │   ├── pt-BR.txt
│   │   ├── ru-RU.txt
│   │   ├── sv-SE.txt
│   │   ├── zh-CN.txt
│   │   └── zh-TW.txt
│   ├── shaders
│   │   ├── applypalette.frag
│   │   ├── applypalette.vert
│   │   ├── applytransparency.frag
│   │   ├── applytransparency.vert
│   │   ├── drawline.frag
│   │   ├── drawline.vert
│   │   ├── drawrect.frag
│   │   └── drawrect.vert
│   └── title
│       ├── openrct2.parkseq
│       ├── rct1aall.parkseq
│       ├── rct1aa.parkseq
│       ├── rct1.parkseq
│       └── rct2.parkseq
├── doc
│   ├── changelog.txt
│   ├── contributors.md
│   ├── licence.txt
│   └── readme.txt
├── libopenrct2.so
├── openrct2
└── openrct2-cli

5 directories, 42 files
```

This is a listing from Travis-produced archive. It looks correct to me, especially `data/shaders/applytransparency.frag` is in there, which is one of the files introduced in this PR.

+1 for Linux builds then.

janisozaur commented:
@duncanspumpkin @IntelOrca anything else is needed before this is merged?

janisozaur commented:
Based on https://github.com/OpenRCT2/OpenRCT2/pull/6541#issuecomment-339356950 and @IntelOrca's approval I'm going to merge this now.

janisozaur commented:
I think this deserves a changelog entry. @LRFLEW can you add one?
